keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+to stop the automation components from acting on other nodes.
+
+### Implementation Details/Notes/Constraints
+
+#### Impact to the Kubelet
+
+None. The kubelet does not need to respond to this lease or interact with it in
+any way.  This is for other components to coordination actions they intend to
+take (or prevent) on the node that might be disruptive.
+
+#### Impact to Scheduleability
+
+None. The existence of acquiring of this lease does not imply any taints,
+annotations, labels, cordoning or draining will take place. The Lease object
+itself is not intended as a mechanism to start any automated process.
+
+A user/controller acquires the Lease object, and then that user/controller
+may add/remove taints, cordon or drain, if they choose to.  Other components
+may wish to integrate with the new Lease object to avoid conflict. EG:
+a component that ensures a taint is always present on a node might stop ensuring
+that taint if there is a current valid lease holder, but that implementation
+is out of scope here.
+
+#### Client Behavior
+
+1. Client has required RBAC permissions for new namespace and Lease object
+1. Client checks for the existence of a lease (to be backwards compatible)
+1. If the lease exists, confirm the lease is not currently held by another user
+1. Attempt to acquire the lease and set an appropriate LeaseDurationSeconds
+1. If acquired, perform the necessary operations
+1. Release the lease
+
+### Risks and Mitigations
+
+Clients might not behave appropriately.  We can mitigate this by providing
+a library with an easy to use interface.
+
+## Alternatives
+
+### Annotations on Node or elsewhere
+
+Use a designated annotation on the node object, or possibly
+elsewhere.  Drawbacks to this include excess updates being applied to the node
+object directly, which might have performance implications. Having such an
+annotation natively seems like an anti-pattern and could easily be disrupted
+by existing controllers/clients attempting to manage those annotations.
+
+### Field in Node Spec Or Status
+
+Extending the node API will be much more difficult.  This behavior is also
+somewhat orthogonal to the node itself.
+
+### Use existing system namespace for lease object
+
+We could utilize an existing system namespace for the lease object. The primary
+issue with this is name collision, especially in the kube-node-lease namespace.
+The kube-node-lease namespace utilizes the same 1-to-1 ratio of Lease objects to
+nodes and also shares node names.  Other existing namespaces might have
+undesirable RBAC requirements.
+
+### Use optional non-system namespace
+
+Administrators that wish to enable this functionality could create a namespace
+with a specific name.  Clients could use a Get-or-Create method to check the
+lease.  This will make installing components that support this
+functionality more difficult as RBAC will not be deterministic at install time
+of those components.
```

Comment:
Another alternative to consider... (I think - I apologize if it doesn't actually meet the use case...) :)

Could we implement this entire workflow outside of core Kubernetes using CRD and custom controllers? I.e. for those who wished to use `MaintenanceLease`, they could deploy the custom controller and create the CRD. I _think_ that within the controller implementation, the proper locking behavior could be implemented.

Reply:
Certainly it's an alternative.  It presents more work for everyone.  If you want to use the CRD based-method, you still need to agree on a namespace (or do system-wide), RBAC, and then introduce a hard-dependency on your project (this CRD needs to be installed first).

I think utilizing the built-in coordination API over a CRD would be more optimal.  We could do things like add an option to kubectl to check for a lease before drain/cordon.

Reply:
I like the idea of an add-on, whether that's using a CRD or an annotation on Node resources. If the core of Kubernetes doesn't provide enough to let the add-on hook in the way this KEP intends, how about providing just enough extension to make the add-on possible?

kubectl could be aware of the add-on and react differently depending on whether a maintenance lease was set.

Reply:
I'm against using a CRD for this.  There is already the coordination API, which seems specifically designed for this type of thing.  I would be okay with an annotation on the node.

I'm unsure what you mean by 'add-on' in this context.

Reply:
I'm suggesting that this node maintenance lease _not_ become part of the Kubernetes core, and instead it becomes a component that a cluster operator can add if they like it: an add-on.

Another KEP, #1308, covers a SIG Cluster Lifecycle mechanism for a cluster operator to configure which add-ons are added into a new cluster, and covers selection of the default configuration as well.

Reply:
@sftim

> I'm suggesting that this node maintenance lease not become part of the Kubernetes core

Why are you suggesting this?  Why would an add-on be preferential here over making it part of the core?